src/main/scala/edg_ide/runner/PDFGeneratorUtil.scala

```diff
@@ -1,41 +1,95 @@
 package edg_ide.runner
 
-import com.lowagie.text.{Document, Element, HeaderFooter, Rectangle}
-import com.lowagie.text.pdf.PdfWriter
+import com.lowagie.text.{Anchor, Document, Element, Font, FontFactory, HeaderFooter, Paragraph, Phrase, Rectangle}
+import com.lowagie.text.pdf.{PdfPCell, PdfPTable, PdfWriter}
+import edg.EdgirUtils.SimpleLibraryPath
 import edg.wir.ProtoUtil.BlockProtoToSeqMap
 import edgir.elem.elem.{BlockLike, HierarchyBlock}
 import org.eclipse.elk.graph.ElkNode
 import edg.wir.DesignPath
 import edg_ide.edgir_graph.HierarchyGraphElk.PropertyMapper
 import edg_ide.edgir_graph.{EdgeWrapper, HierarchyGraphElk, NodeDataWrapper, PortWrapper}
 import edg_ide.swing.blocks.ElkNodePainter
+import edgir.ref.ref.LibraryPath
 
 import java.awt.Color
 import java.io.{FileNotFoundException, FileOutputStream, IOException}
-
+import scala.collection.mutable
 
 object PDFGeneratorUtil{
 
+  /*
+    Note for future development:
+      If infinite loop (and eventually StackOverflow) occurs when adding header/footer, check that the
+      page size is large enough/page margin is small enough so that header/footer content do not exceed
+      the available space in the document.
+   */
+  final val PAGE_MARGIN = 15
+  final val font = new Font(Font.HELVETICA, 12f, Font.UNDERLINE, Color.BLUE)
+  final val TABLE_ROW_HEIGHT = 23f
+
+
   private def generatePageSize(node: ElkNode): (Float, Float) = {
-    val width = node.getWidth.toFloat + 2 * ElkNodePainter.margin.toFloat
-    val height = node.getHeight.toFloat + 2 * ElkNodePainter.margin.toFloat
+    val width = node.getWidth.toFloat + 2.5f * ElkNodePainter.margin.toFloat
+    val height = node.getHeight.toFloat + 2.5f * ElkNodePainter.margin.toFloat
     (width, height)
   }
 
+
+  def getDuplicationList(block: HierarchyBlock, path: DesignPath = DesignPath(),
```

Comment:
Can you make this functional instead? That is, instead of it taking in a mutable.Map and mutating it in-place, it just returns a filled Map? Having it be functional eliminates potential bugs where it's passed in a non-empty map or confusion about what's supposed to be passed in.

I see it's recursive, there's two broad ways you could do this:
Option 1: the function takes the result of the recursive call(s) and merges them, returning the merged result
Option 2: the function internally uses mutable data structures (using inner functions that modify that mutable structure), but externally does not depend on mutable structures

Additionally, consider adding a short comment / summary about what this does functionally.

Reply:
Will look at changing this!